File: go/src/binarytree/binarytree.go
package main

import (
	"fmt"
	"os"
)

type Node struct {
	left, right *Node
	value int
}

func (n *Node) Add(v int) {
	if n.value == -1 {
		n.value = v
	} else if v < n.value {
		if n.left == nil {
			n.left = &Node{nil, nil, v}
		} else {
			n.left.Add(v)
		}
	} else {
		if n.right == nil {
			n.right = &Node{nil, nil, v}
		} else {
			n.right.Add(v)
		}
	}
}

var listindex int = 0
var list [1000]int

func (n *Node) Prefix() {
	if n.left != nil {
		n.left.Prefix()
	}
	list[listindex] = n.value
	listindex++
	if n.right != nil {
		n.right.Prefix()
	}
}

func main() {
	root := &Node{nil, nil, -1}
	N := 500
	iterations := 1000
	listindex = 0

	for iterations > 0 {
		iterations--
		root.Add(N)
		N = (2 * N + 10) % 1000
	}

	root.Prefix()
	if (listindex < 1000) {
		os.Exit(1)
	}
	for i := 1; i < 1000; i++ {
		if (list[i] < list[i - 1]) {
			fmt.Println("OOPS")
			os.Exit(i)
		}
	}
}

File: cpp/binarytree.cpp
#include <iostream>
#include <vector>

class Node {
public:
	Node(int v) : value(v) { }
	void add(int v);
	void prefix(std::vector<int>&);
private:
	int value;
	Node *left = nullptr;
	Node *right = nullptr;
};

void Node::add(int v) {
	if (v <= this->value) {
		if (!this->left) {
			this->left = new Node(v);
		} else {
			this->left->add(v);
		}
	} else {
		if (!this->right) {
			this->right = new Node(v);
		} else {
			this->right->add(v);
		}
	}
}

void Node::prefix(std::vector<int>& vec) {
	if (this->left) {
		this->left->prefix(vec);
	}
	vec.push_back(this->value);
	if (this->right) {
		this->right->prefix(vec);
	}
}

int main() {
	int N = 500;
	const int iterations = 1000;
	std::vector<int> vec;

	Node root(N);

	for (int i=0; i<iterations; ++i) {
		N = (2 * N + 10) % 1000;
		root.add(N);
	}

	root.prefix(vec);
	if (vec.size() < 1000) {
		return 1;
	}

	for (int i = 1; i < iterations; i++) {
		if (vec[i] < vec[i - 1]) {
			return i;
		}
	}

	return 0;
}
